src/sinks/file/mod.rs

```diff
@@ -58,6 +58,11 @@ pub struct FileSinkConfig {
     )]
     pub compression: Compression,
 
+    /// Precise quality based on the underlying compression algorithms’ qualities.
+    /// The interpretation of this depends on the algorithm chosen and the specific implementation
+    /// backing it. Qualities are implicitly clamped to the algorithm’s maximum.
+    pub compression_level: Option<u32>,
```

Comment:
I think I'd like to see this as a field on `enum Compression` below if possible, something like:
```rust
enum Compression {
    Gzip {
        level: Option<u32>,
    },
    Zstd {
        level: Option<u32>,
    },
    None,
}
```
I think serde's "adjacently tagged" representation might get us to backwards compatibility with existing configs, but a custom deserializer might be necessary.

Reply:
Good point. I will try to implement locally. If I am reading serde documentation right, it should be much better. Thanks!

Reply:
@bruceg by the way, do you know a right way to put the documentation from `compression_level` field to the `level` field inside `Compression` enum?

Reply:
I think you can just use the normal doc comments (`///`) at all levels.

Reply:
Yeah. the only caveat in this case is the comment duplicatoin. Not so critical in this case, I guess.

Reply:
@bruceg I have uploaded the implementation based on adjacent tagging. Unfortunately, it does not work and I cannot understand, what is wrong.

For the following config:
```
[sources.demo]
type = "file"
include = [ "example_copy.txt" ]
data_dir = "/Users/zamazan4ik/open_source/vector/target/debug"
read_from = "beginning"
ignore_checkpoints = true

[sinks.file]
type = "file"
inputs = ["demo"]
compression = "zstd"
compression_level = 1
path = "/Users/zamazan4ik/open_source/vector/target/debug/vector-%Y-%m-%d.zst"

  [sinks.file.encoding]
  codec = "json"
```

I have the following error:
```
2022-09-10T00:10:59.213573Z  INFO vector::app: Log level is enabled. level="vector=info,codec=info,vrl=info,file_source=info,tower_limit=trace,rdkafka=info,buffers=info,kube=info"
2022-09-10T00:10:59.215149Z  INFO vector::app: Loading configs. paths=["vector.toml"]
2022-09-10T00:10:59.222585Z ERROR vector::cli: Configuration error. error=invalid type: string "zstd", expected adjacently tagged enum Compression for key `sinks.file`
```

I am not a serde expert, so I need a piece of advice here, how can I fix it.

Reply:
I think you have two choices here (both of them requiring flattening):
```rust
#[serde(tag = "compression")] // internally tagged
enum Compression {
    Gzip { compression_level: Option<u32>, },
    …
}

#[serde(tag = "compression", content = "compression_level")] // adjacently tagged
enum Compression {
    Gzip(Option<u32>),
    …
}
```

What I would prefer as a longer-term solution would be to allow configuration either of the following:
```toml
compression = "zstd" # shorthand for compression.method

compression.method = "zstd"
compression.level = 9
```

That would require a custom deserializer, or at least a wrapper, which could be a fair bit trickier to write.

Reply:
@bruceg Unfortunately, I still did not get a working solution even with flattening (that was a surprise for me that serde flatten does not work on enums, only on structs).

For the following file:
```
[sources.demo]
type = "file"
include = [ "example_copy.txt" ]
data_dir = "/Users/zamazan4ik/open_source/vector/target/debug"
read_from = "beginning"
ignore_checkpoints = true

[sinks.file]
type = "file"
inputs = ["demo"]
compression = "zstd"
compression_level = 1
path = "/Users/zamazan4ik/open_source/vector/target/debug/vector-%Y-%m-%d.zst"

  [sinks.file.encoding]
  codec = "json"
```

With the current commit I have the following error: 
```
2022-09-18T01:10:02.619608Z  INFO vector::app: Log level is enabled. level="vector=info,codec=info,vrl=info,file_source=info,tower_limit=trace,rdkafka=info,buffers=info,kube=info"
2022-09-18T01:10:02.621331Z  INFO vector::app: Loading configs. paths=["vector.toml"]
2022-09-18T01:10:02.628681Z ERROR vector::cli: Configuration error. error=invalid type: string "zstd", expected internally tagged enum Compression for key `sinks.file`
```

I have also tried use `serdr(flatten)` directly on `compression` field in `FileSinkConfig` - also without success.

Reply:
Oh, I did not realize flatten wouldn't work with enums. 🤔 I don't want to hang up what is otherwise a very simple change on this configuration issue, but it would be good to move forward into a clean UI for it. I think the best alternative would be a struct, something like:

```rust
struct Compression {
    mode: CompressionMode,
    level: Option<usize>,
}

impl From<String> for Compression {
    fn from(mode: String) -> Self {
        Self { mode, level: None }
    }
}

enum CompressionMode {
    Zstd, etc
}

struct Config {
    …
    #[serde(deserialize_with = "vector_core::serde::string_or_struct")]
    compression: Compression,
}
```

This `string_or_struct` doesn't exist yet, but is almost a straight copy-and-paste from `bool_or_struct` that is there.

If you want, I can make the changes here to your branch, or you can continue to work on this. Let me know how you want to proceed.

Reply:
@bruceg Sorry for the so long response. Unfortunately, I will be busy with other work. Could you please commit directly in this branch? It will help a lot to land the change sooner.

Thanks in advance!